Fix regression in visibility of `st.code`'s copy-to-clipboard button

snehankekre opened the pull request:
## 📚 Context

Fixes #6490. PR #6042, while refactoring `st.code` to be markdown independent, likely introduced a regression in `st.code`, where the copy button does not appear when you hover over code blocks. This issue was caused by missing CSS hover rules and the way the hover effect was implemented. The button is now restored and appears on hover as expected.

- What kind of change does this PR introduce?

  - [x] Bugfix
  - [ ] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

- Modified `StreamlitSyntaxHighlighter.tsx `to include the `StyledCopyButtonContainer` and `StyledCodeBlock` components, which were previously missing.
- Updated the `StyledCopyButtonContainer` in `styled-components.ts` to set the correct opacity, width, and pointerEvents properties, ensuring that the copy button is correctly rendered for `st.code`.
- Added the necessary hover and active states to `StyledCopyButton` in `styled-components.ts`.
- Removed the redundant `StyledCopyButtonContainer` import and usage from `StreamlitMarkdown.tsx`, as it's now integrated into `StreamlitSyntaxHighlighter`.

  - [ ] This is a breaking API change
  - [x] This is a visible (user-facing) change

**Revised:**

![image](https://user-images.githubusercontent.com/20672874/232417749-6488b64a-d524-48ce-a61c-2b8dc4a279f5.png)

**Current:**

![image](https://user-images.githubusercontent.com/20672874/232417780-60bfc4fd-586e-41fc-b5ff-071cf1b629be.png)

## 🧪 Testing Done

- [x] Screenshots included
- [ ] Added/Updated unit tests
- [ ] Added/Updated e2e tests

## 🌐 References

_Does this depend on other work, documents, or tickets?_

- **Issue**: Closes #6490 

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

snehankekre committed:
Fix regression in `st.code` copy button and update `ElementNodeRenderer`

snehankekre commented:
TODO: write e2e test such that Cypress hovers over the code block (so the copy button is visible) and then takes snapshots.

sfc-gh-kbregula commented:
We have two usages of the `StreamlitSyntaxHighlighter` component.
https://github.com/streamlit/streamlit/blob/4289ea809619804043705318c1aef9795b57f2f1/frontend/src/components/core/Block/ElementNodeRenderer.tsx#L640-L645
https://github.com/streamlit/streamlit/blob/4289ea809619804043705318c1aef9795b57f2f1/frontend/src/components/shared/StreamlitMarkdown/StreamlitMarkdown.tsx#L273-L275
In both cases we need these buttons, so I think we can move this code inside the `StreamlitSyntaxHighlighter` component to avoid code duplication.

snehankekre committed:
Move copy button rendering and its container styles from ElementNodeRenderer.tsx and StreamlitMarkdown.tsx to StreamlitSyntaxHighlighter.tsx

snehankekre commented:
> In both cases we need these buttons, so I think we can move this code inside the StreamlitSyntaxHighlighter component to avoid code duplication.

Thanks for that insight, @sfc-gh-kbregula! Fixed on 42a1a3a. 

There was also a bug in my initial fix: for `st.code`, the copy button appears only when you hover over the extreme right side of the code block, whereas with `st.markdown`, it appears when you hover anywhere in the code block. Also fixed by 42a1a3a:

- Moved the copy button rendering and its container styles from `ElementNodeRenderer.tsx` and `StreamlitMarkdown.tsx` to StreamlitSyntaxHighlighter.tsx. This centralizes the copy button handling in one place, removing code duplication.
- Updated `StyledCopyButtonContainer` and `StyledCopyButton` styles in `styled-components.ts` for better positioning and visibility. The `StyledCopyButtonContainer` now has pointerEvents: "none" and the `StyledCopyButton` has pointerEvents: "auto" to ensure proper click handling.
- Updated the `StreamlitSyntaxHighlighter` component to render the copy button and its container when the provided children prop is a string. This ensures the copy button is correctly displayed for both `st.code` and `st.markdown`.

snehankekre committed:
Add a conditional check to only render copy button when code block is not empty

sfc-gh-kbregula commented:
@snehankekre Do you need help with tests? I'll be happy to deal with it.

snehankekre committed:
Remove the extra vertical space between code blocks caused by the added StyledCodeBlock

snehankekre commented:
@sfc-gh-kbregula that would be awesome, thank you! 🙏 Reading the Cypress docs on how to hover and then write tests would take me a couple hours.. time I could instead spend on `st.connection` docs if possible.

snehankekre committed:
fix: adjust spacing between code blocks in StyledCodeBlock

sfc-gh-kbregula committed:
Add E2E tests

sfc-gh-kbregula commented:
@snehankekre I added E2E tests.

sfc-gh-kbregula committed:
fixup! Add E2E tests

snehankekre commented:
LGTM, thanks Kamil!

sfc-gh-kbregula commented:
@snehankekre Do you need any more help with this PR? Is there anything left to do?

snehankekre commented:
All that's left for this PR is code review.

sfc-gh-kbregula commented on frontend/src/components/elements/CodeBlock/styled-components.ts:
```diff
@@ -105,50 +105,63 @@ export const StyledPre = styled.pre(({ theme }) => ({
+    color: theme.colors.bodyText,
+  },
+
+  "&:active, &:focus, &:hover": {
```
I wonder why, we have two code blocks with identical styles.

snehankekre replied:
~Which two code blocks are you referring to?~

snehankekre replied:
- `StyledPre` is responsible for the styling of the code itself, including various tokens and colors.
- `StyledCopyButtonContainer` sets up the container for the copy button, positioning it correctly within the code block and handling hover states.
- `StyledCodeBlock` provides the necessary relative positioning and margins for the code block.
- `StyledCopyButton` is the styling for the copy button itself, handling its appearance and hover states (sets the position, size, color, and various styles to show and hide the button on hover and click).

Each of these styled components serves a unique purpose IIUC

snehankekre replied:
Ahh, I see what you mean! I can simplify the styles for `StyledCopyButton` merging the individual selectors into one string with a comma separator. 

On aff7539, I apply the same styles to the `StyledCopyButton` when either its parent `StyledCodeBlock` is being hovered or when the `StyledCopyButton` itself is in an active, focused, or hovered state

snehankekre committed:
Refactor `StyledCopyButton` styles to remove redundancies
